src/nlsolve/type.jl

```diff
@@ -1,72 +1,87 @@
+@enum NLStatus::Int8 begin
+  Convergence = 1
+  TryAgain = 0
+  Divergence = -1
+end
+
 # solver
 
 abstract type AbstractNLSolver{algType,iip} end
 
-mutable struct NLSolver{algType,iip,uType,uTolType,tTypeNoUnits,C<:AbstractNLSolverCache} <: AbstractNLSolver{algType,iip}
+mutable struct NLSolver{algType,iip,uType,tType,C<:AbstractNLSolverCache} <: AbstractNLSolver{algType,iip}
   z::uType
   tmp::uType
   ztmp::uType
-  γ::uTolType
-  c::tTypeNoUnits
+  γ::tType
+  c::tType
   alg::algType
-  κ::uTolType
-  fast_convergence_cutoff::uTolType
-  ηold::uTolType
+  κ::tType
+  fast_convergence_cutoff::tType
+  ηold::tType
   iter::Int
   maxiters::Int
   status::NLStatus
   cache::C
 end
 
+NLSolver{iip,tType}(z, tmp, ztmp, γ, c, alg, κ, fast_convergence_cutoff,
+                    ηold, iter, maxiters, status, cache) where {iip,tType} =
+  NLSolver{typeof(alg), iip, typeof(z), tType, typeof(cache)}(z, tmp, ztmp, convert(tType, γ),
+                                                              convert(tType, c), alg, convert(tType, κ),
+                                                              convert(tType, fast_convergence_cutoff),
+                                                              convert(tType, ηold), iter,
+                                                              maxiters, status, cache)
+
 # caches
 
-mutable struct NLNewtonCache{uType,tType,rateType,uNoUnitsType,J,W,du1Type,ufType,jcType,lsType,G} <: AbstractNLSolverCache
+mutable struct NLNewtonCache{uType,tType,rateType,J,W,ufType,jcType,lsType} <: AbstractNLSolverCache
   ustep::uType
   tstep::tType
   k::rateType
-  atmp::uNoUnitsType
+  atmp::uType
   dz::uType
   J::J
   W::W
   new_W::Bool
   W_dt::tType
-  du1::du1Type
+  du1::uType
   uf::ufType
   jac_config::jcType
   linsolve::lsType
   weight::uType
-  invγdt::G
+  invγdt::tType
   new_W_dt_cutoff::tType
+  J_t::tType
 end
 
-mutable struct NLNewtonConstantCache{tType,J,W,ufType,G} <: AbstractNLSolverCache
+mutable struct NLNewtonConstantCache{tType,J,W,ufType} <: AbstractNLSolverCache
   tstep::tType
   J::J
   W::W
   new_W::Bool
   W_dt::tType
   uf::ufType
-  invγdt::G
+  invγdt::tType
   new_W_dt_cutoff::tType
 end
 
-mutable struct NLFunctionalCache{uType,tType,rateType,uNoUnitsType} <: AbstractNLSolverCache
+mutable struct NLFunctionalCache{uType,tType,rateType} <: AbstractNLSolverCache
   ustep::uType
   tstep::tType
   k::rateType
-  atmp::uNoUnitsType
+  atmp::uType
```

Comment:
It seems this change (and similar ones in the other caches) will break the support for number types with units?

Reply:
Yes, that is intentional. We plan to drop units support for implicit solvers, since linear solvers cannot solve matrices with units. Also, having a state vector with homogeneous units isn't particularly interesting either. CC @ChrisRackauckas

Reply:
yeah, nonlinear solving with units just ain't happening any time soon. When all of the other tooling gets in place, we can swap that out again. For now, we can make things a little simpler for the next 5 years...